src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala

```diff
@@ -0,0 +1,138 @@
+package lms.transformation.tensor
+
+import scala.annotation.implicitNotFound
+import scala.collection._
+
+import lms.core._
+import lms.core.stub._
+import lms.collection.mutable._
+import lms.macros.SourceContext
+import lms.thirdparty.array_computation.{ArrayCPUOps, CUDATypeLess, CudaOps}
+
+import Backend._
+
+trait FixedSizeDistributedTensorBinaryTypeLess extends FixedSizeDistributedTensorMutationTypeLess {
+
+  def ElemWiseNoBroadCasting(x: TENSOR, y: TENSOR, anno: Anno, __pos: SourceContext)(op: String): TENSOR = {
+    assert(x.shape_size == y.shape_size)
+    assert(x.et == y.et)
+    (new TENSOR(Adapter.g.reflectRead(op, C(x.tensor_type), C(anno), x.x, y.x)(x.x, y.x))).withSrcType(__pos, x.et)
+  }
+
+  def Add(x: TENSOR, y: TENSOR, anno: Anno = NAnno)(implicit __pos: SourceContext): TENSOR =
+    ElemWiseNoBroadCasting(x, y, anno, __pos)("tensor_add")
+
+  def Sub(x: TENSOR, y: TENSOR, anno: Anno = NAnno)(implicit __pos: SourceContext): TENSOR =
+    ElemWiseNoBroadCasting(x, y, anno, __pos)("tensor_minus")
+
+  def Mul(x: TENSOR, y: TENSOR, anno: Anno = NAnno)(implicit __pos: SourceContext): TENSOR =
+    ElemWiseNoBroadCasting(x, y, anno, __pos)("tensor_mult")
+
+  def Div(x: TENSOR, y: TENSOR, anno: Anno = NAnno)(implicit __pos: SourceContext): TENSOR =
+    ElemWiseNoBroadCasting(x, y, anno, __pos)("tensor_div")
+
+  override def mergable_dims(node: Node) = node match {
+    case Node(s, op, tt::anno::(x:Backend.Sym)::(y:Backend.Sym)::_, _)
+        if (op == "tensor_add" || op == "tensor_minus" || op == "tensor_mult" || op == "tensor_div") =>
+      val x_type = (new TENSOR(x, useOldMetadata=true)).tensor_type
+      val y_type = (new TENSOR(y, useOldMetadata=true)).tensor_type
+      (x_type.shape zip y_type.shape).toList map { case (a:Size, b:Size) => (a.dim, b.dim) }
+    case _ => super.mergable_dims(node)
+  }
+
+  override def aircopCollect(node: Node, forwardNodes: mutable.ArrayBuffer[Node],
+      weightNodes: mutable.ArrayBuffer[Node], backwardNodes: mutable.ArrayBuffer[()=>Unit],
+      grad_map: mutable.HashMap[Backend.Sym, TENSOR],
+      momentum_map: mutable.HashMap[Backend.Sym, TENSOR],
```

Comment:
Nit: `forwardNodes`/`backwardNodes` vs `grad_map`/`momentum_map`, better to use a unified naming style

Reply:
will fix